File: quickstart-python/services/jwt_service.py
import jwt
import requests

from consts import SecretKeys, EnvironmentKeys
from errors import GenericUnauthorizedError, InternalServerError
from services import SecretService, EnvironmentVariablesService


class JWTService:
    @staticmethod
    async def create_jwt_token(payload: dict):
        """
        Create a JWT token with the given payload.
        """
        signing_secret = await SecretService.get_secret(SecretKeys.MONDAY_SIGNING_SECRET)
        return jwt.encode(payload, signing_secret, algorithm='HS256')

    @staticmethod
    async def decode_monday_jwt(token: str, verify=True):
        """
        Validate a JWT token.
        """
        try:
            signing_secret = await SecretService.get_secret(SecretKeys.MONDAY_SIGNING_SECRET)
            if not signing_secret:
                raise InternalServerError('No signing secret found')
            return jwt.decode(token, signing_secret,
                              options={"verify_aud": False, "verify_signature": verify}, algorithms=['HS256'])
        except jwt.ExpiredSignatureError:
            raise GenericUnauthorizedError('Token has expired')
        except jwt.InvalidSignatureError:
            raise GenericUnauthorizedError(
                'Invalid signature, token validation failed')
        except jwt.InvalidTokenError:
            raise GenericUnauthorizedError('Invalid token')

    @staticmethod
    async def get_monday_token(code):
        """
        Function that exchanges a code for a monday.com token.
        """
        token_path = await EnvironmentVariablesService.get_environment_variable(EnvironmentKeys.MONDAY_OAUTH_TOKEN_PATH)
        client_id = await EnvironmentVariablesService.get_environment_variable(EnvironmentKeys.MONDAY_OAUTH_CLIENT_ID)
        client_secret = await SecretService.get_secret(SecretKeys.MONDAY_OAUTH_CLIENT_SECRET)

        response = requests.post(token_path, data={
            'code': code,
            'client_id': client_id,
            'client_secret': client_secret
        })
        response.raise_for_status()  # Raises a HTTPError if the response status is 4xx, 5xx
        return response.json()

File: quickstart-node/src/services/jwt.service.ts
import { EnvironmentVariablesService, SecretsService } from '@services/monday-code';
import jwt, { JsonWebTokenError, JwtPayload, TokenExpiredError } from 'jsonwebtoken';
import { InternalServerError, Unauthorized } from 'http-errors';
import axios from 'axios';
import { EnvironmentsKeys, SecretsKeys } from '@shared/monday-auth-consts';

export class JWTService {
  private constructor() {}

  static createJWTToken(payload: object): string {
    const signingSecret = SecretsService.getInstance().get(SecretsKeys.MONDAY_SIGNING_SECRET) as string;
    return jwt.sign(payload, signingSecret, { algorithm: 'HS256' });
  }

  static decodeMondayJWT(token: string, verify: boolean = true): JwtPayload {
    const signingSecret = SecretsService.getInstance().get(SecretsKeys.MONDAY_SIGNING_SECRET) as string;
    if (!signingSecret) {
      throw new InternalServerError('No signing secret found');
    }
    try {
      return (
        verify
          ? jwt.verify(token, signingSecret, {
              algorithms: ['HS256'],
            })
          : jwt.decode(token)
      ) as JwtPayload;
    } catch (error: any) {
      if (error instanceof TokenExpiredError) {
        throw new Unauthorized('Token has expired');
      }
      if (error instanceof JsonWebTokenError) {
        throw new Unauthorized('Invalid signature, token validation failed');
      }
      throw new Unauthorized('Invalid token');
    }
  }

  static async getMondayToken(code: string): Promise<any> {
    const tokenPath = EnvironmentVariablesService.getInstance().get(EnvironmentsKeys.MONDAY_OAUTH_TOKEN_PATH) as string;
    const clientId = EnvironmentVariablesService.getInstance().get(EnvironmentsKeys.MONDAY_OAUTH_CLIENT_ID);
    const clientSecret = SecretsService.getInstance().get(SecretsKeys.MONDAY_OAUTH_CLIENT_SECRET);

    const response = await axios.post(tokenPath, {
      code: code,
      client_id: clientId,
      client_secret: clientSecret,
    });

    return response.data;
  }
}
